### PR TITLE
Refactor Firebase Admin Stubbing

### PR DESCRIPTION
**Pull request by samtstern**
### Description

Fixes issue #1454
	 
This investigation took me very deep and along the way I also:

  * Added some standardization for how we resolve developer node modules and made sure we log them.
  * Simplified some of our logging code.

### Scenarios Tested

Using these functions to test:

```js
const functions = require("firebase-functions");
const admin = require('firebase-admin');

const app = admin.initializeApp();

exports.writeThroughAdmin = functions.https.onRequest(async (req, res) => {
    const firestore = admin.firestore();
    const ref = firestore.doc("info/test");
    await ref.set({ hello: (new Date().toISOString()) });
    ref.get().then((snap) => {
        res.json({ data: snap.data() });
    });
});

exports.writeThroughApp = functions.https.onRequest(async (req, res) => {
    const firestore = app.firestore();
    const ref = firestore.doc("info/test");
    await ref.set({ hello: (new Date().toISOString()) });
    ref.get().then((snap) => {
        res.json({ data: snap.data() });
    });
});

exports.firestoreReaction = functions.firestore.document("info/test").onWrite(async (change, ctx) => {
    console.log('~~~~~~~~~~ FIRESTORE WRITE ~~~~~~~~~~');
    return true;
});
```

I also added a unit test that verifies the same behavior.

### Sample Commands

N/A

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -230,6 +230,7 @@
           "resolved": "https://registry.npmjs.org/gaxios/-/gaxios-1.8.4.tgz",
```
Comment (samtstern) [inline package-lock.json]: Note: these changes should disappear once we get `master` fixed.

File: src/emulator/functionsEmulatorRuntime.ts [hunk 2]
```diff
@@ -137,56 +135,79 @@ class Proxied<T> {
...
+    return obj !== undefined;
+  }
 }
 
 async function resolveDeveloperNodeModule(
   frb: FunctionsRuntimeBundle,
-  pkg: PackageJSON,
   name: string
 ): Promise<ModuleResolution> {
+  // TODO: Can we cache this call?  Should we?
```
Comment (samtstern) [inline src/emulator/functionsEmulatorRuntime.ts]: I should cache this before merge.

**Comment by samtstern** [comment]
@yuchenshi thanks for taking a look.  I agree with waiting for @abeisgoat

File: src/emulator/functionsEmulatorRuntime.ts [hunk 3]
```diff
@@ -492,105 +511,138 @@ async function getGRPCInsecureCredential(frb: FunctionsRuntimeBundle): Promise<a
...
-  new EmulatorLog("DEBUG", "runtime-status", "firebase-admin has been stubbed.", {
+  logDebug("firebase-admin has been stubbed.", {
     adminResolution,
-  }).log();
+  });
+
   return adminModuleProxy;
 }
 
+function proxyFirebaseApp(
```
Comment (abeisgoat) [inline src/emulator/functionsEmulatorRuntime.ts]: Are there any changes to this logic?

Comment (samtstern) [reply src/emulator/functionsEmulatorRuntime.ts]: Yes.  The logic changed such that `app.firestore()` and `admin.firestore()` go to the same place.  PTAL.

**Pull request merged**